### DISCUSSION TITLE
Inlined local function

### DISCUSSION THREAD

**Discussion by teneko** [dotnet/csharplang#7346] [Language Ideas] [answered]
We know there exists local/nested functions and lambda/anonymous functions. Here I only talk about local/nested functions!

Has it ever been discussed to define something like this:

```csharp
void InvokeWriteHelloWorld(Action<string> write) =>
  write("Hello World");

void WriteHelloWorldToConsole() {
  InvokeWriteHelloWorld(WriteToConsole);

  void WriteToConsole(string content) =>
    Console.WriteLine(content);
}
```

as:

```csharp
void InvokeWriteHelloWorld(Action<string> write) =>
  write("Hello World");

void WriteHelloWorldToConsole() {
  InvokeWriteHelloWorld(WriteToConsole(string content) => Console.WriteLine(content));
}

// or long form

//void WriteHelloWorldToConsole() {
//  InvokeWriteHelloWorld(WriteToConsole(string content) {
//     Console.WriteLine(content);
//   });
//}
```

where `WriteToConsole` is treated as named local function and the return type is inferred. This would be a kind of QoL-update where we easily write local functions in lambda-like-fashion but inlined/not required to write it explicitly.

E.g.:

```csharp
abstract void AcceptEnumerator(Func<IAsyncEnumerable<object>> enumerableFactory);

AcceptEnumerator(YieldReturnObjects() {
  yield return new object();
});
```

instead of:

```csharp
abstract void AcceptEnumerator(Func<IAsyncEnumerable<object>> enumerableFactory);

IAsyncEnumerable<object> YieldReturnObjects() {
  yield return new object();
}

AcceptEnumerator(YieldReturnObjects);
```

**Comment by svick** [comment]
I don't understand, for the first case why is the existing lambda syntax not sufficient? And the second one is odd, because it looks like you want to implicitly invoke the declared function? So I think you should clarify what exactly you want, and why do you think it would be a good idea. (And why having names that are never used anywhere is necessary.)

If you want to use `yield return` in a lambda, then there is an existing suggestion for that: https://github.com/dotnet/csharplang/discussions/4601. But https://github.com/dotnet/roslyn/issues/24#issuecomment-70448545 does not make it seem particularly likely.

**Reply by teneko** [reply]
Maybe I don't find the right words, because I am not that in touch with the language features as you, so please excuse me.

My understanding is, that lambda is fully supported to be represented as AST. Therefore any new C# language (e.g. state machines) are not immediatelly or never ported to lambda. I can also imagine, that the anonymous functions need some adoption to new C# features too. Local functions on the other hand are like any other function and capable to be nested to static functions and instance methods. Therefore they also do support the "yield (return/break)"-functionality and the magic behind it automatically, right?

So currently, to pass a local function handle as parameter, you always need to define the function explicitly first. I would like to shorten this. That is what I wanted to demonstrate with the second example. This is rather a quality of life thing than anything else.

So to your last statement, yes having native support for `yield return` in lambdas would be "cool", but using local functions have their advantages too as outlined in https://github.com/dotnet/roslyn/issues/24#issuecomment-70448545, but as the last statement of it also summarizes:

> Not perfect, but better than creating a delegate object just to immediately invoke and discard it.

**Reply by CyrusNajmabadi** [reply]
> My understanding is, that lambda is fully supported to be represented as AST

Not sure what this means.

> Therefore any new C# language (e.g. state machines) are not immediatelly or never ported to lambda. 

This is not true.  Lambdas can be async, for example.  We haven't done iterators just because we haven't found it important enough.

**Reply by teneko** [reply]
Ah please be nice. I am really not into the technical details of all the great features C# provides.

> We haven't done iterators just because we haven't found it important enough.

So you didn't ported iterators to lambda. I think I really tried to not be to offensive when trying to explain that new features need to be ported to the lambda syntax, am I right? Please correct me. 

My thinking is really simple. We have currently two ways to build "anonymous" functions. local functions and anonymous functions, but only local functions support everything and anonymous functions for example (and the only thing what I am aware of) are not capable of iterators. So I would make the usage of local functions more user-friendly to make many nested local functions not a hell of block inside a block inside a block. Example:

```csharp
abstract object WrapEnuemrator(Func<IEnumerable<object>> enumerableFactory);

IEnumerable<object> Level0()
{
    yield return WrapEnuemrator(Level1);

    IEnumerable<object> Level1()
    {
        yield return WrapEnuemrator(Level2);

        IEnumerable<object> Level2()
        {
            yield return WrapEnuemrator(Level3);

            IEnumerable<object> Level3()
            {
                yield return new object();
            }
        }
    }
}
```

versus

```csharp
abstract object WrapEnuemrator(Func<IEnumerable<object>> enumerableFactory);

IEnumerable<object> Level0()
{
    yield return WrapEnuemrator(Level1()
    {
        yield return WrapEnuemrator(Level2()
        {
            yield return WrapEnuemrator(Level3()
            {
                yield return new object();
            });
        });
    });
}
```

**Reply by CyrusNajmabadi** [reply]
The general point i'm making is: we currently have found the use case of 'iterators' in lambdas important enough to do.

So using the case of 'iterators' to motivate being able to use local-functions as expression falls under the same scrutiny.  

We'd need *Additional* motivation to change the thinking here :)

Thanks!

**Reply by teneko** [reply]
Maybe a named lambdas would be more natural to us C-Sharpler, but would be then treated as local function. Somehting like this:

```csharp
abstract object WrapEnuemrator(Func<IEnumerable<object>> enumerableFactory);

IEnumerable<object> Level0()
{
    yield return WrapEnuemrator(Level1() =>
    {
        yield return WrapEnuemrator(Level2() =>
        {
            yield return WrapEnuemrator(Level3() =>
            {
                yield return new object();
            });
        });
    });
}
```

**Reply by CyrusNajmabadi** [reply]
> Maybe a named lambdas would be more natural to us C

This is still just a request to allow iterators in lambdas :)  THat's def a reasonable request, but our thinking still is that this is a narrow enough case that isn't rising to the level of something we want to invest in :)  sorry!

**Reply by teneko** [reply]
@CyrusNajmabadi I see and fully understand! But my point is not to support iterators in lambdas, but make inlined local functions maybe an option or an alternative, to make nested local functions as proposed a little more user-friendly in the scenario I tried to outline above.

**Reply by teneko** [reply]
Otherwise I would have taken another title i.e. "Allow iterators in lambdas", but I have chosen "Inlined local function", so I really try to only communicate to this feature where iterators is just a detail in the discussion, haha. I hope you get me now. =)

**Reply by CyrusNajmabadi** [reply]
>  but make inlined local functions maybe an option or an alternative

What benefit would this bring though?  What's the motivation for us to do that when lambdas *already* exist?

Basically, for us to do this, we need to good use cases to help support why this would be worth doing.  Thanks!

**Reply by svick** [reply]
> My understanding is, that lambda is fully supported to be represented as AST. Therefore any new C# language (e.g. state machines) are not immediatelly or never ported to lambda.

If you're talking about expression trees, then those are indeed limited and generally don't get new language features. E.g. `Expression e = (object o) => o is 42;` produces an error: 

> error CS8122: An expression tree may not contain an 'is' pattern-matching operator.

But you're not using expression trees, so you *can* use new language features like pattern matching. Iterator blocks (`yield return`) are the major exception to that. But I don't see a reason to introduce a brand new syntax for that.

---

> But my point is not to support iterators in lambdas, but make inlined local functions maybe an option or an alternative, to make nested local functions as proposed a little more user-friendly in the scenario I tried to outline above.

What makes the proposed syntax more "user-friendly"? What I see is a more verbose syntax for lambdas, with unnecessary name. (And also the ability to use `yield return`, but you said that's not the point.)

**Reply by teneko** [reply]
> What benefit would this bring though? What's the motivation for us to do that when lambdas already exist?

You already are telling the thing. There exist two different concepts. There is one reason for lambdas and one reason for local functions. Now we have local functions but you are telling me, that local functions ARE lambdas (exaggerated), but they are not. Now we have the case, that local functions always need to be defined explictly before their handles can be passed as parameters. So why in the first place you introduced the short syntax `blubb(() => {})` to express an inlined function? Because you do not wanted to first write `Action nothing = new Action(() => {})` and pass that down to `blubb(nothing)`.

So now I would like to have something very similiar to local functions, where local functions can be similiar described as `blubb(() => {})` for example as `blubb(nothing() => {})`, but maybe `nothing() => {}` could be treated as local function as one idea. And I am here to let you know my motivation to this, to have a short form, the same reason why we can now write `void WriteToConsole() => Console.WriteLine();` instead of `void WriteToConsole() { Console.WriteLine(); }`.

So I really try to be neutral and optimistic. I really would be happy when you are not immediatelly against a new proposal because you read lambdas and iterators in one sentence.

**Reply by teneko** [reply]
You are right, the name would be kinda redundant, but then we would have a real lambda, but we would have to differentiate between a real lambda and still a local function. The name was an initial idea, to give the compiler a clue how he should name the local function.

**Reply by CyrusNajmabadi** [reply]
> There is one reason for lambdas and one reason for local functions

The primary reason for local functions is simply to be a method *without* an allocated delegate.  But in all the cases you've shown, there would be an allocated delegate.  So it woudl be the same as a lambda.

> where local functions can be similiar described as `blubb(() => {})` for example as `blubb(nothing() => {})`

I get the request :)   The point i'm making is that the user doesn't seem to get any value of that form *over* just writing `blubb(() => {})`.   So what i'm asking is: What is hte *benefit* to doing this?  Why would a user use the latter form over the former one?

Thanks!

> I really would be happy when you are not immediatelly against a new proposal

We're not immediately against it.  I'm simply stating that we need a compelling reason to do this.  Right now it just looks like an alternative, more verbose, form of a lambda.  :)

**Reply by svick** [reply]
>  And I am here to let you know my motivation to this, to have a short form

But you already have a short form, it's `() => {}`. What is the advantage of the proposed syntax over that? What does being "treated as local function" mean and how would it affect the code?

> So I really try to be neutral and optimistic. I really would be happy when you are not immediatelly against a new proposal because you read lambdas and iterators in one sentence.

The C# language is quite deliberate in what features it adds, because each feature has significant costs, both in the resources that are needed to add it and in making the language larger. Which means that the vast majority of features suggested here will never be implemented and that suggestions receive a lot of scrutiny, which can be perceived as negativity by the people who propose features.

**Reply by CyrusNajmabadi** [reply]
>  but then we would have a real lambda, but we would have to differentiate between a real lambda and still a local function. 

I think you need to explain this further.  What's the difference for you between a "Real lambda" and a "local function"?  In both cases we're just allocating a delegate instance that, when executed, will run the body code provided.  :)

Fundamentally, a local-function is interesting when it is *not* passed as a value to something, but is instead directly invoked (so it does not become a delegate instance).  In that case, there is no allocation, and there is just direct method invocation. :)

**Reply by teneko** [reply]
First of all, we discuss this in an already very deep way. So to recap my view of things. I am consumer and see two concepts. I see local functions as any other instance method or static functions. At the end of the day, so my understanding, can all instance methods be seen as as static functions. But lambdas, to my knowledge are hard-coded, part of the body code of a method and always allocated at time where the anonymous delegate is needed (wow hard to explain!). After reading what you have written I get what you mean: If you pass the a local function as parameter, it behaves like a lambda, because now the concept of delegate are required: for example having variable lifetime scope of parent method block if the local function is not static and so on, right?

So now we have two ways to allocate a delegate, one by passing a local function and one by passing an anonymous function - got it. So after having a better understanding I understand why you ask now for the benefit of `example() => {}`. And to be honest, I do not know it. My initial idea was treat a lambda not as an anonymous function, but as an local function that has been just inlined. This would be one option to allow (as side effect) iterators à la `example() => { yield return new object(); }`. But again, since both representation, on the one side the local function and on the other side the lambda functions, are equivalent at the moment when you pass it as parameter, it is more or less not "correct" to introduce a verbose form like `example() => {}`. But mhh. So at the end of the day lambdas have to support iterators, right?

**Reply by teneko** [reply]
So the only benefit would be to make use of any new and existing C# features in anonymous function (`() => {}`), when this anonymous function could be somehow seen as local function from the view of the compiler. The name was one idea to make the difference between the underlying lambda (syntax) implementation and the existing local function functionality. Now I see, that this is not a good approach. It would be interesting when you could introduce a syntax like `apply(function () => {})` to somehow tell the compiler to treat the anonymous function as a local function. The only other possiblity is to let lambda learn iterators and all upcoming features that will maybe come which seems a little bit odd, when lambda cannot be that expressive as local functions.

**Reply by HaloFour** [reply]
The "verbose" form suggested here doesn't "fix" the problem with anonymous iterators in the form of lambdas.  The lack of the name was never the problem.  In fact, there really is no technical problem preventing anonymous iterators, just a lack of interest in implementing it in the language.

**Reply by HaloFour** [reply]
To clarify, the reason you can't do `yield` in a lambda today has nothing to do with syntax.  It has everything to do with a lack of interest.

**Reply by teneko** [reply]
Can you eloborate this? Whats the problem with this?

```
abstract void DoSomethingWithBlubb(Func<IEnumerable<object>> blubbHandle);

// This is not an anonymous iterator, just a very normal iterator written as local function
IEnumerable<object> Blubb() {
  yield return new object();
}

Func<IEnumerable<object>> blubbHandle = Blubb;
DoSomethingWithBlubb(Blubb);
```

Why would you now need to reimplement this in the language?

So in my head is the following mental model: local function > lambda > delegate (allocation)

And I would now be able to treat a lambda as local function to make use of all functionality of a local function inside a lambda function; something like:

```
abstract void DoSomethingWithBlubb(Func<IEnumerable<object>> blubbHandle);

Func<IEnumerable<object>> blubbHandle = function () => { // prepend "function" for example to indicate local function
  yield return new object();
};
DoSomethingWithBlubb(blubbHandle);

// or 

DoSomethingWithBlubb(function () => {
  yield return new object();
});
```

In the background the lambda would would have been translated to a local function and not seen anymore as lambda, because you "annotated" the lambda as local function, so the compiler can happily create the magic classes for the state machine of the iterator functions. I hope you can still follow my weird thoughts.

**Reply by CyrusNajmabadi** [reply, marked as answer]
As mentioned several times, the use case of using an iterator in a direct delegate value is *not enough of* a motivating case for us. :-)

So if that's the only benefit here for having this new syntax for a lambda/local-function, it's going to be similarly not motivating enough. 

We need more of a reason to do this. 

Can you give a *different* benefit?  Something else that isn't possible today that this would enable our help with?  

For example, as I mentioned, async/await works with both. `static` is also available for both. Etc. Etc.

**Reply by teneko** [reply]
As I am only one consumer (of the C# language), and have faced this problematic while outweighting the idea to implement a library around iterators. I have no more benefits for the one or the other proposal that has been discussed here.

The thing I take away is that the usage of iterators won't be improved in the near future, so I would not consider to implement a library around iterators at all, because from consumer view it is not attractive to implement dozen of nested local functions in such an explicitly manner. I think your view is absolutely understandable when considering that you need to choose wisely what to introduce and what not.

So last question: since this does not get considered and iterators won't probably never be supported by lamdas, is there any chance to contribute to the latter feature? I am absolutely not familiar with the technical side of the language, but I always wanted to contribute to the langauge I write my code for a decade now.

**Reply by CyrusNajmabadi** [reply]
> is there any chance to contribute to the latter feature?

Just so we're on the same page. What are your referring to with "latter feature" in that sentence?  It wasn't clear to me from context. Thanks!

Wrt contributing to the language, any issue we have in the "any time" milestone is one I Believe we would work with the community on having a community implemented solution. Thanks!

**Reply by teneko** [reply]
With the latter I mean iterators in anonymous functions you paused because there was not enough incentive.

**Reply by CyrusNajmabadi** [reply]
Yes. We would only accept this as a community implemented feature if there is a proposal in the "any time" millstone

**Reply by teneko** [reply]
In the future I really would like to port something like Kotlin Coroutines or Redux Saga to C#, but sadly the language is not yet ready for this.

**Reply by CyrusNajmabadi** [reply]
I haven't directly used react-saga, but I'm familiar with kotlin.  I personally prefer the more opinionated and library-extended form that we have. You're welcome to prefer their style, but we haven't found our approach leading to problematic gaps in practice.

**Reply by teneko** [reply]
I really talk about Kotlin Coroutines that makes use of suspend functions that are equivalent to yield return and not Kotlin at all, and I do not know, but react-saga does not exist. I really mean Redux Saga, where you write generators (similiar to iterators but where yield return can actually be assigned to a variable) to write orchestrations in a fancy way. In this area C# is definitely lacking of features. No critic I still love C#. <3

**Reply by CyrusNajmabadi** [reply]
Right. I'm familiar. My preference is the built in machinery of async/await. I haven't found a case in tons of real world coding where the more generic form would provide enough value.

**Reply by CyrusNajmabadi** [reply]
> but react-saga does not exist. I really mean Redux Saga,

Sorry. Fat-fingered.  I meant the latter :-)

**Reply by teneko** [reply]
Indeed, the async await functionality is great, but implementing orchestrations in state machines can be useful too. In C# you can actually do it, but it is very verbose and explicit and reducing the boilerplate not possible because of language limitations: yield return not assignable, errors cannot be bubbled up to the iterator and iterators have to be written as local functions. still everything can be workaround to such a degree that no one would actually use the library. :D

**Comment by HaloFour** [comment]
Why wouldn't you just use lambdas in these cases?

I'm not sure I understand the second case.  It seems you want to pass the return value of the local function, but as a method group?

**Reply by teneko** [reply]
Sorry the second case was hand-written and has been fixed. Would you take another look?

**Reply by teneko** [reply]
To have a better understanding, I try to explain it further.

In the code, where the local function `YieldReturnObjects` has been inlined:

```csharp
abstract void AcceptEnumerator(Func<IAsyncEnumerable<object>> enumerableFactory);

AcceptEnumerator(YieldReturnObjects() {
  yield return new object();
});
```

is this:

```csharp
YieldReturnObjects() {
  yield return new object();
}
```

the local function. As a quick idea I left away the return type, because it could be inferred, right?

So the lambda version would be:

```csharp
AcceptEnumerator(() => {
  yield return new object(); // Error CS1621 The yield statement cannot be used inside an anonymous method or lambda expression
})
```

and the inlined local function version could be:

```csharp
AcceptEnumerator(DescriptiveName() {
  yield return new object();
})
```

The compiler would autoamtically interpret it as:

```csharp
IAsyncEnumerable<object> DescriptiveName() {
  yield return new object();
}

AcceptEnumerator(DescriptiveName)
```

**Reply by teneko** [reply]
But I have to say that this syntax is not bullet-proof and can be easily misinterpreted as you did I think. The nearest example I have in mind is Kotlin's inlined functions that you can call similiar to the syntax I suggest: https://kotlinlang.org/docs/inline-functions.html

**Reply by FaustVX** [reply]
So what you need is that the lambda function to have the full potential of regular function (eg: `yield`)
Because, what's the point to have a named function inlined ? You won't call it anywhere else.

**Reply by teneko** [reply]
Yes. It would be more or less only for shorter syntax to allow someone as library author of for example coroutines to make the heavy usage of iterators a blessing and not a hell. And because I didn't find any infos to this, I tried to communicate this here. =)
